File: backend/app/services/code_review_levels.py
"""Experience-level tuning for AI code reviews."""

from typing import Literal

ExperienceLevel = Literal["beginner", "intermediate", "advanced", "senior"]

DEFAULT_EXPERIENCE_LEVEL: ExperienceLevel = "intermediate"

EXPERIENCE_LEVEL_LABELS: dict[ExperienceLevel, str] = {
    "beginner": "Beginner",
    "intermediate": "Intermediate",
    "advanced": "Advanced",
    "senior": "Senior",
}

_LEVEL_INSTRUCTIONS: dict[ExperienceLevel, str] = {
    "beginner": """The developer is a BEGINNER (early career or learning to code).
- Use plain language; briefly explain *why* something matters when you mention a concept.
- Prioritize fundamentals: project structure, readable naming, basic error handling, and obvious bugs.
- Be encouraging — call out what they did well before suggesting fixes.
- Avoid heavy jargon; when you use a term (e.g. "DRY", "separation of concerns"), add a one-line explanation.
- Focus on 3–5 high-impact, achievable improvements rather than exhaustive nitpicks.
- Severity: reserve "critical/high" for issues that could break the app or leak data; frame others as learning opportunities.""",
    "intermediate": """The developer is at an INTERMEDIATE level (1–3 years, comfortable shipping features).
- Balance teaching with professional standards — assume they know basics but benefit from best practices.
- Cover code quality, testing gaps, security basics, and maintainability.
- Suggestions should be concrete with brief rationale; reference common patterns and tools where helpful.
- Highlight both strengths and a prioritized improvement list.
- Include moderate-depth notes on architecture only when clearly relevant to the codebase size.""",
    "advanced": """The developer is ADVANCED (experienced engineer, owns features or modules end-to-end).
- Assume strong fundamentals; focus on architecture, design patterns, scalability, performance, and edge cases.
- Discuss trade-offs explicitly (e.g. simplicity vs flexibility).
- Call out test strategy, observability, error boundaries, and API design where applicable.
- Fewer "info" items — prioritize high-signal findings.
- Suggestions can reference industry patterns, refactoring strategies, and production readiness.""",
    "senior": """The developer is SENIOR/STAFF level (leads technical direction, mentors others).
- Review as a peer staff/principal engineer: systems thinking, long-term maintainability, and organizational impact.
- Emphasize architectural boundaries, failure modes, security posture, operability, and technical debt trade-offs.
- Challenge assumptions; discuss alternatives and when current choices are acceptable vs when to refactor.
- Minimal hand-holding — concise, precise language; assume deep context.
- Priorities should reflect strategic impact (reliability, velocity, risk), not style nits.""",
}


def normalize_experience_level(level: str | None) -> ExperienceLevel:
    if level in EXPERIENCE_LEVEL_LABELS:
        return level  # type: ignore[return-value]
    return DEFAULT_EXPERIENCE_LEVEL


def build_review_system_prompt(level: ExperienceLevel | str | None = None) -> str:
    """Return the system prompt tuned for the developer's experience level."""
    normalized = normalize_experience_level(level if isinstance(level, str) else level)
    label = EXPERIENCE_LEVEL_LABELS[normalized]
    level_block = _LEVEL_INSTRUCTIONS[normalized]

    return f"""You are an AI software engineer performing a code review tailored to a {label}-level developer.

AUDIENCE & TONE:
{level_block}

Analyze the project structure, code quality, security, architecture, testing, dependencies, and documentation.

Respond with ONLY valid JSON (no markdown fences) matching this schema:
{{
  "overall_score": 0-100,
  "summary": "2-4 sentence executive summary appropriate for this experience level",
  "strengths": ["string", ...],
  "priorities": ["top 3-5 actionable improvements ordered by impact for this level"],
  "categories": [
    {{
      "name": "Architecture|Security|Code Quality|Testing|Performance|Documentation|Dependencies",
      "score": 0-100,
      "findings": [
        {{
          "severity": "critical|high|medium|low|info",
          "file": "path or null",
          "line": null or number,
          "title": "short title",
          "description": "what you found",
          "suggestion": "concrete fix"
        }}
      ]
    }}
  ]
}}

Be specific — reference actual files when possible. Calibrate depth and vocabulary to the {label} level.
If the codebase looks solid, still provide info-level suggestions appropriate for this level."""

File: frontend/src/lib/codeReviewLevels.ts
/** Code review experience levels — tunes AI review depth and tone. */

export type ExperienceLevel = "beginner" | "intermediate" | "advanced" | "senior";

export const CODE_REVIEW_LEVEL_KEY = "nexusai-code-review-level";

export const EXPERIENCE_LEVELS: {
  value: ExperienceLevel;
  label: string;
  description: string;
}[] = [
  {
    value: "beginner",
    label: "Beginner",
    description: "Friendly explanations, fundamentals, and encouraging feedback",
  },
  {
    value: "intermediate",
    label: "Intermediate",
    description: "Balanced professional review with best practices",
  },
  {
    value: "advanced",
    label: "Advanced",
    description: "Architecture, patterns, performance, and edge cases",
  },
  {
    value: "senior",
    label: "Senior",
    description: "Staff-level systems thinking and strategic trade-offs",
  },
];

export function loadExperienceLevel(): ExperienceLevel {
  const stored = localStorage.getItem(CODE_REVIEW_LEVEL_KEY);
  if (
    stored === "beginner" ||
    stored === "intermediate" ||
    stored === "advanced" ||
    stored === "senior"
  ) {
    return stored;
  }
  return "intermediate";
}

export function saveExperienceLevel(level: ExperienceLevel): void {
  localStorage.setItem(CODE_REVIEW_LEVEL_KEY, level);
}

export function experienceLevelLabel(level: ExperienceLevel): string {
  return EXPERIENCE_LEVELS.find((l) => l.value === level)?.label ?? level;
}
